### DISCUSSION TITLE
HC-SR04 Distance Sensor

### DISCUSSION THREAD

**Discussion by MercuryKira** [gpiozero/gpiozero#1002] [Q&A]
Operating system: Raspbian Buster
Python version: 3.7.3
Pi model: Pi 4 Model B
GPIO Zero version: 1.6.2
Pin factory used: RPiGPIO
See http://rpf.io/gpzissue for information on how to find out these details

Please give us as much information about your issue as possible. Write your
code inside code blocks like so:

from example code:
```python
from gpiozero import DistanceSensor
from time import sleep

sensor = DistanceSensor(echo=17, trigger=27)
while True:
    print('Distance: ', sensor.distance * 100)
    sleep(1)
```
show eror:
/usr/lib/python3/dist-packages/gpiozero/input_devices.py:856: PWMSoftwareFallback: For more accurate readings, use the pigpio pin factory.See https://gpiozero.readthedocs.io/en/stable/api_input.html#distancesensor-hc-sr04 for more info
  'For more accurate readings, use the pigpio pin factory.'
/usr/lib/python3/dist-packages/gpiozero/input_devices.py:948: DistanceSensorNoEcho: echo pin set high
  warnings.warn(DistanceSensorNoEcho('echo pin set high'))

but my code work
```python
import RPi.GPIO as GPIO
import time
GPIO.setmode(GPIO.BCM)

class DisSensor():
    def __init__(self, Triger=17, Echo=27):
        self.Triger = Triger
        self.Echo = Echo
        GPIO.setup(self.Triger, GPIO.OUT)
        GPIO.setup(self.Echo, GPIO.IN)
    
    def distance(self):
        GPIO.output(self.Triger, 1)
        time.sleep(0.00001)
        GPIO.output(self.Triger, 0)
        StartTime = 0
        StopTime = 0
        while GPIO.input(self.Echo) == 0:
            StartTime = time.time()

        while GPIO.input(self.Echo) == 1:
            StopTime = time.time()

        TimeElapsed = StopTime - StartTime
        
        Temp = 20
        SoundSpeed=(331.5+0.6*Temp) 
        
        distance = (TimeElapsed*SoundSpeed*100)/2
        
        return distance
    
if __name__ == '__main__':
    sens = DisSensor()
    k=0
    while True:
        dis = sens.distance()
        k+=1
        print(k,"Measured Distance = %.1f cm" % dis)
        time.sleep(0.30)
```

**Comment by bennuttall** [comment]
Those are warnings, not errors.

The first says that you'll get better accuracy by switching to the pigpio pin factory. See https://gpiozero.readthedocs.io/en/stable/api_pins.html#changing-the-pin-factory

The second says that it didn't receive an echo back. If this happens all the time, even when objects are directly in front of the sensor, then maybe something's wrong. If it only happens once, then it should still be working properly.

1. Try using the pigpio pin factory
2. Make sure you don't have any other programs running which might be using the same pins
3. Make sure you test it with an object directly in front of the sensor, and see if you reliably get a reading out of it

**Reply by MercuryKira** [reply]
```python
from gpiozero.pins.pigpio import PiGPIOFactory
from gpiozero import DistanceSensor
from time import sleep

sensor = DistanceSensor(echo=17, trigger=27, pin_factory = PiGPIOFactory())
while True:
    print('Distance: ', sensor.distance * 100)
    sleep(1)
```
The object is 7 cm away from it, and the code above shows this. But now it just throws an warning and does not show the distance:
/usr/lib/python3/dist-packages/gpiozero/input_devices.py:978: DistanceSensorNoEcho: no echo received
  warnings.warn(DistanceSensorNoEcho('no echo received'))

**Reply by lurch** [reply]
Perhaps it's because one is using:
```python
sensor = DistanceSensor(echo=17, trigger=27)
```
but the other is using:
```python
def __init__(self, Triger=17, Echo=27):
```
i.e. you appear to have swapped the pin-numbers around?

**Reply by bennuttall** [reply]
Good spot!

**Comment by MercuryKira** [comment]
Everything was soldered and it worked. I tried to poll the pins in the code anyway.
